Drivers OTF backend

jterry64 opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help -->

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [X] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [X] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [X] Check the commit's or even all commits' message styles matches our requested structure.
- [X] Check your code additions will fail neither code linting checks nor unit test.



## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: GTC-3167


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

- Now call OTF query service to compute results in a background task from POST
- POST immediately stores initial "pending" status in file and GET checks that for whether to use Retry-After
- Compute background task stores results in a file
- GET requests read results from file
- Change tests to use mocks instead of full end-to-end

## Does this introduce a breaking change?

- [X] Yes
- [ ] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

Slight changes to the model (mostly the metadata) of things I realized while working on it. This will be updated in the ticket GTC-3114.

## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

jterry64 committed:
Use OTF backend for drivers instead of precanned response

jterry64 committed:
Use mocks for tree cover loss by drivers tests

jterry64 committed:
Use pydantic for tcl by driver compute task

jterry64 committed:
Use more consistent naming

jterry64 committed:
Deal with error from OTF in drivers background task

jterry64 committed:
Add logging to background task

jterry64 committed:
Check geostore is valid on POST

jterry64 committed:
Fix error writing

danscales commented on app/routes/datamart/land.py:
```diff
@@ -131,20 +84,33 @@ async def tree_cover_loss_by_driver_get(
 )
 async def tree_cover_loss_by_driver_post(
     data: TreeCoverLossByDriverIn,
+    background_tasks: BackgroundTasks,
```
Is the deprecated=True at line 83 intended, or maybe just a copy/paste error from somewhere else?

jterry64 replied:
oh, copy paste error

jterry64 replied:
@danscales removed

danscales commented on app/routes/datamart/land.py:
```diff
@@ -131,20 +84,33 @@ async def tree_cover_loss_by_driver_get(
 
+    # check geostore is valid
+    try:
+        await get_geostore(data.geostore_id, GeostoreOrigin.rw)
```
Not sure if this is worth fixing, but this get_geostore() call access checks both rw and gfw geostores, so it will succeed for a gfw geostore id.  But in compute_tree_cover_loss_by_driver, the get_geostore() calls requires assumes the geostore is from RW, will fail if it is from GFW, I think.

jterry64 replied:
Oh, I may have misunderstood this, I thought this would also check gfw too. I want to check both, will double check how we normally do that.

danscales replied:
Actually, ignore this comment, I got confused.  You're using the same get_geostore() call in both, so you will check both geostores in both cases, and there's no inconsistency.

jterry64 committed:
Fix test

jterry64 committed:
Add version overrides option for OTF query

jterry64 committed:
Add tests for dataset overrides

jterry64 committed:
Remove deprecation flag

dmannarino commented on app/routes/datamart/land.py:
```diff
@@ -66,85 +73,62 @@ async def tree_cover_loss_by_driver_get(
     # trigger background task to create item
     # return 202 accepted
-    resource_id = _get_resource_id(data.geostore_id, data.canopy_cover)
+    dataset_version = DEFAULT_LAND_DATASET_VERSIONS | data.dataset_version
```
At first I thought you were using a default OR the one provided, and thought you had it the wrong way round. But this is actually merging two dictionaries, right?

jterry64 replied:
Yes! New syntax in Python 3.9+ I just learned about, I think I got the ordering right based on my test results... basically merges two dicts, with the one on the right overwriting any clashing keys

jterry64 commented on app/routes/datamart/land.py:
```diff
@@ -66,85 +73,62 @@ async def tree_cover_loss_by_driver_get(
+        status_code=200,
+        headers=headers,
+        content={"data": resource, "status": "success"},
+    )
```
I think response model would make sense. In fact, I think we already have some JSEND wrapper I should be using?

jterry64 committed:
Add decorator to make dict hasheable in _get_data_environment

jterry64 committed:
Commit decorator module

jterry64 committed:
Fixed broken cache_clear() due to new decorator on _get_data_environment

jterry64 committed:
Truly use pydantic models everywhere

jterry64 commented:
@danscales @dmannarino I think I resolved all your comments if you want to take a quick look again. Main change is I switched to using consistent pydantic models everywhere, and also added the ability to specify a dataset/version.

jterry64 committed:
Merge branch 'develop' into gtc-3167/drivers-otf-backend

danscales commented on app/models/orm/queries/raster_assets.py:
```diff
@@ -1,7 +1,4 @@
-from ...enum.pixetl import Grid
-
-sql = """
+latest_raster_tile_sets = """
```
Can you maybe change the name of this variable, since it is now all versions of all raster tile sets (on the right grid), rather than just the latest versions?

danscales commented on app/models/pydantic/datamart.py:
```diff
@@ -1,8 +1,49 @@
+class DataMartResource(StrictBaseModel):
+    status: str = "saved"
+    details: Optional[str] = None
+    metadata: DataMartMetadata = None
```
Change DataMartMetadata -> Optional[DataMartMetadata] to make typing more correct.

danscales commented on app/routes/datasets/queries.py:
```diff
@@ -449,7 +459,9 @@ async def _query_dataset_json(
         geostore = cast(GeostoreCommon, geostore)
-        results = await _query_raster(dataset, default_asset, sql, geostore)
+        results = await _query_raster(
+            dataset, default_asset, sql, geostore, raster_version_overrides
```
**You need to specify "version_overrides=raster_version_overrides", since there are other omitted parameters format and delimiter before version_overrides.

danscales commented on app/routes/datasets/queries.py:
```diff
@@ -811,10 +827,14 @@ def _get_default_layer(dataset, pixel_meaning):
     # get all raster tile set assets with the same grid.
-    latest_tile_sets = await db.all(latest_raster_tile_sets, {"grid": grid})
+    sql = _get_data_environment_sql(version_overrides)
+    latest_tile_sets = await db.all(db.text(sql), {"grid": grid})
```
Would be nice to change 'latest_tile_sets' to 'matching_tile_sets', since they are no longer just latest tile versions.

danscales reviewed (approved):
Looks good, except for the one case of wrong args for _query_raster() call.  Other comments are minor.

Nice to be able to specify versions other than the latest!

jterry64 committed:
Use snake case consistently for drivers endpoint

jterry64 committed:
Fix PR comments

jterry64 commented:
@danscales thanks for the comments, resolved them and will merge the PR at the end of the day if you're fine with my resolutions

jterry64 committed:
Merge branch 'develop' into gtc-3167/drivers-otf-backend
